nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -818,35 +822,57 @@ impl DataStore {
         opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
         let conn = self.pool_connection_authorized(opctx).await?;
 
-        let result = paginated(sled_dsl::sled, sled_dsl::id, pagparams)
+        // We're going to build the query in stages.
+        //
+        // First, select all active sleds, and join the `sled` table with the
+        // `vmm` table on the VMM's sled_id`, filtering out VMMs which are not
+        // actually incarnated on a sled.
+        let query = sled_dsl::sled
+            .inner_join(vmm_dsl::vmm.on(vmm_dsl::sled_id.eq(sled_dsl::id)));
+        // Next, paginate the results, ordering by the sled ID first, so that we
+        // list all VMMs on a sled before moving on to the next one, and then by
+        // the VMM ID.
+        //
+        // Note that we must add the `paginated_multicolumn` wrapper
+        // at this point in query construction, because here, the selection
+        // contains both `sled_dsl::id` and `vmm_dsl::id` columns, but it does
+        // *not* have anything that makes it no longer implement
+        // `diesel::QuerySource`, which the `paginated_multicolumn` function
+        // requires. This ordering doesn't actually matter when it comes to the
+        // generated SQL, which should be equivalent no matter how we construct
+        // the query, but it *does* matter for satisfying Diesel's trait
+        // constraints.
+        let query = paginated_multicolumn(
+            query,
+            (sled_dsl::id, vmm_dsl::id),
+            pagparams,
+        );
+        // Filter out sleds that aren't in service, and VMMs that aren't
+        // incarnated on a sled.
+        let query = query
             .filter(sled_dsl::time_deleted.is_null())
             .sled_filter(SledFilter::InService)
```

Comment:
Are these two redundant?

Reply:
I don't believe so --- IIUC expunging a sled does not immediately delete its database record?